documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+
+    public TScalar LengthSquared { get; }
+    public TScalar Length();
+
+    public Vector2F<TScalar> Dot(Vector2F<TScalar> other);
+    public TScalar DistanceSquared(Vector2F<TScalar> other);
+    public TScalar Distance(Vector2F<TScalar> other);
+    public Vector2F<TScalar> Abs();
+    public void CopyTo(TScalar[] array);
+    public void CopyTo(TScalar[] array, int index);
+
+    public static Vector2F<TScalar> Dot (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> DistanceSquared (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Distance (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Clamp (Vector2F<TScalar> value, Vector2F<TScalar> min, Vector2F<TScalar> max);
+    public static Vector2F<TScalar> Min (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Max (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Abs (Vector2F<TScalar> value);
+}
+```
+At the end of this proposal is the definition for the new interfaces used in the these classes.
+
+# Matrix Types
+
+This proposal includes the following matrix types:
+- Matrix2x2F
+- Matrix2x2I
+- Matrix2x3F
+- Matrix2x3I
+- Matrix2x4F
+- Matrix2x4I
+- Matrix3x2F
+- Matrix3x2I
+- Matrix3x3F
+- Matrix3x3I
+- Matrix3x4F
+- Matrix3x4I
+- Matrix4x2F
+- Matrix4x2I
+- Matrix4x3F
+- Matrix4x3I
+- Matrix4x4F
+- Matrix4x4I
+- Matrix5x4F
+- Matrix5x4I
+
+Integer Variants do not require any functions which interact with Quaternions
+
+Matricies **must** fulfill the following requirements:
+- be a struct
+- Stored in row major format
+- F matricies work with F vectors, and I Matricies work with I vectors
+- Both column and row vectors and individual values accessible via properties
+- An indexer that takes row and column indicies and outputs the value
+- Add, subtract, and multiply operators defined with Matricies of the same size
+- Multiply operators defined with compatible matricies, if the output matrix type already exists (AxB * BxC = AxC)
+- Multiply operator defined with the appropriate Vector type (vector4F * Matrix4x4F is valid)
+  - **optionally**, include multiplication operators for Vector types smaller than the appropriate type, assuming 1 for the missing dimensions (vector2F * Matrix4x4F -> (Vector2F.x, Vector2F.y, 1, 1) * Matrix4x4F)
+- Negate Operator defined
+- Fulfills `IEquatable<T>` where `T` is the same matrix class and the System.Numerics version of the matrix if available
+- Implicit conversion to and from the System.Numerics matrix type, if available
+- Invert function for square matricies
+- GetDeterminant function for square matricies and Matrix3x2
+- Transpose function
+- static lerp function
+- static identity property
+- For Matrix3x2, Matrix3x3, Matrix4x3, and Matrix4x4 include the following static functions
+  - CreateBillboard
```

Comment:
Some of these should probably have left and right handed versions.

System.Numerics is modeled after XNA, which used "right handed" by default. IIRC this is also what GLM uses.
DirectX uses left handed by default

Beyond left vs right handed, there is then the consideration on whether Z is up or Y is up (a lot of games make Z up, so that top down works for 2D and the coordinates make sense when displayed in UI) and whether the screen's top left is 0, 0 or the bottom left is 0, 0. (others are possible too)

These other nuances can sometimes influence what people think about when doing various operations, but GLM and DX do agree themselves on how values work for left vs right handed, so Silk should match that.